models/block.go

```diff
@@ -28,17 +29,34 @@ func NewBlockFromBytes(data []byte) (*Block, error) {
 	var b *Block
 	err := rlp.DecodeBytes(data, &b)
 	if err != nil {
-		return nil, err
+		pastBlock := decodeBlockBreakingChanges(data)
+		if pastBlock == nil {
+			return nil, err
+		}
+		h, err := pastBlockHash(pastBlock)
+		if err != nil {
+			return nil, err
+		}
+		b = pastBlock
+		b.hash = &h
 	}
 
 	return b, nil
 }
 
 type Block struct {
 	*types.Block
+	hash              *gethCommon.Hash
 	TransactionHashes []gethCommon.Hash
 }
 
+func (b *Block) Hash() (gethCommon.Hash, error) {
+	if b.hash != nil {
+		return *b.hash, nil
+	}
+	return b.Block.Hash()
+}
```

Comment:
Do we need the caching ?

Reply:
Is this Block object serialized? if yes are we worried about caching the hash?

Reply:
This Block object is indeed serialized, but the `hash` field is only set for past blocks. Where past block means a block that fails to decode with the current format.